metadata/src/utils/retain.rs

```diff
@@ -5,187 +5,132 @@
 //! Utility functions to generate a subset of the metadata.
 
 use crate::{
-    ExtrinsicMetadata, Metadata, OuterEnumsMetadata, PalletMetadataInner, RuntimeApiMetadataInner,
-    StorageEntryType,
+    ExtrinsicMetadata, Metadata, PalletMetadataInner, RuntimeApiMetadataInner, StorageEntryType,
+};
+use alloc::collections::{BTreeSet, VecDeque};
+use scale_info::{
+    PortableType, TypeDef, TypeDefArray, TypeDefBitSequence, TypeDefCompact, TypeDefComposite,
+    TypeDefSequence, TypeDefTuple, TypeDefVariant,
 };
-use alloc::collections::BTreeMap;
-use hashbrown::HashSet;
-use scale_info::TypeDef;
-
-/// Collect all type IDs needed to represent the provided pallet.
-fn collect_pallet_types(pallet: &PalletMetadataInner, type_ids: &mut HashSet<u32>) {
-    if let Some(storage) = &pallet.storage {
-        for entry in storage.entries() {
-            match entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    type_ids.insert(ty);
-                }
-                StorageEntryType::Map {
-                    key_ty, value_ty, ..
-                } => {
-                    type_ids.insert(key_ty);
-                    type_ids.insert(value_ty);
-                }
-            }
-        }
-    }
 
-    if let Some(ty) = pallet.call_ty {
-        type_ids.insert(ty);
-    }
+#[derive(Clone)]
+struct TypeSet {
+    seen_ids: BTreeSet<u32>,
+    work_set: VecDeque<u32>,
+}
 
-    if let Some(ty) = pallet.event_ty {
-        type_ids.insert(ty);
+impl TypeSet {
+    fn new() -> Self {
+        Self {
+            seen_ids: BTreeSet::new(),
+            work_set: VecDeque::with_capacity(22),
+        }
     }
 
-    for constant in pallet.constants.values() {
-        type_ids.insert(constant.ty);
+    fn insert(&mut self, id: u32) -> bool {
+        self.seen_ids.insert(id)
     }
 
-    if let Some(ty) = pallet.error_ty {
-        type_ids.insert(ty);
+    fn contains(&mut self, id: u32) -> bool {
+        self.seen_ids.contains(&id)
     }
-}
 
-/// Update all type IDs of the provided pallet using the new type IDs from the portable registry.
-fn update_pallet_types(pallet: &mut PalletMetadataInner, map_ids: &BTreeMap<u32, u32>) {
-    if let Some(storage) = &mut pallet.storage {
-        for entry in storage.entries.values_mut() {
-            match &mut entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    update_type(ty, map_ids);
-                }
-                StorageEntryType::Map {
-                    key_ty, value_ty, ..
-                } => {
-                    update_type(key_ty, map_ids);
-                    update_type(value_ty, map_ids);
+    /// This function will deeply traverse the inital type and it's dependencies to collect the relevant type_ids
+    fn collect_types(&mut self, metadata: &Metadata, id: u32) {
+        self.work_set.push_back(id);
+        while let Some(typ) = self.work_set.pop_front() {
+            if self.insert(typ) {
```

Comment:
A quick suggestion I oftten apply is to invert the condition to avoid nesting, ie here:

```rust
if !self.insert(ty) {
    // We hit a type we've already inserted; avoid infinite loops and stop.
    continue
}

// Other code now less nested :)
```

Reply:
done :D